### ISSUE TITLE
Lab3 review

### ISSUE THREAD

**Issue by GiorgioBongiovanni** [mistru97/CI2024_lab3#1]
As far as I'm concerned, this implementation of the $n^2-1$ puzzle solver exhibits both strengths and weaknesses.
Firstly, a positive aspect is that the program is modular, separating concerns into distinct functions (for example, available_actions, do_action, goal_test, search): this makes the code easier to follow and maintain but some parts of the code lack comments making it harder to understand.  
Secondly, multiple heuristics are provided (h_euclidean, h_manhattan, h_tile_not_in_place) allowing flexibility in the choice of the heuristic for guiding the search but, based on the tests that I performed, the Manhattan heuristic is the most effective for grid-based puzzles while you chose euclidean distance.
This code implements a graph search, as evidenced by the use of the state_cost dictionary to track the minimum cost for each visited state.
This strategy prevent re-exploration of the same states, which is a hallmark of graph search, unlike tree search which does not account for visited states.
According to:
```python
if new_state not in state_cost and new_state not in [s[1] for s in frontier]:
```  
a new state is considered only if it has not already been visited (not in state_cost) or is not already in the frontier (the check uses a list comprehension, which could be optimized with a set).
Therefore, the graph search requires the heuristic to be both admissible and consistent.
The Manhattan distance is more suitable for grid-based problems where movements are restricted to vertical and horizontal directions because it directly reflects the real constraints of such movements. 
The Manhattan distance between two points measures the total number of steps required to move from one point to another along a grid where only vertical and horizontal moves are allowed and so reflects the actual movement cost in grid-based problems.
In a grid where movements are restricted to vertical and horizontal directions:
- A path from a point $(x_1, y_1)$ to another point $(x_2, y_2)$ must consist of $|x_2 - x_1|$ horizontal steps and $|y_2 - y_1|$ vertical steps.
- The Manhattan distance exactly captures this cost because it counts the required horizontal and vertical steps separately and sums them up.

The Manhattan distance is always admissible in grid-based problems because it represents the minimum possible cost to reach the goal, assuming no obstacles and it is always consistent because moving from state s to the next state s' in one vertical or horizontal step reduces the Manhattan distance by exactly 1, matching the transition cost c(s, s') = 1.
In addition, the Manhattan distance only requires subtraction and addition of coordinates, which are computationally inexpensive operations.
This makes it particularly suitable for problems with large grids or many states to evaluate.

On the other hand, the Euclidean distance between two points is 
$\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}$
The Euclidean distance assumes straight-line movement which includes diagonal moves. 
In grid-based problems where only vertical and horizontal moves are allowed, this does not align with the actual cost of moving between states and thus underestimates too much the true cost.

The Euclidean distance is admissible because it never overestimates the true straight-line distance.  
However, it is not always consistent in grid problems because it does not accurately account for the discrete steps required.
This can lead to situations where the heuristic suggests a lower cost than the actual transition requires violating the consistency condition.

For instance, consider a grid where movements are restricted to vertical and horizontal steps:
**Start**: (0, 0)
**Goal**: (3, 4)  

1. **Manhattan Distance**:
   - Required steps: $|3 - 0| + |4 - 0| = 3 + 4 = 7$

2. **Euclidean Distance**:
   - Straight-line distance:  
     $\sqrt{(3 - 0)^2 + (4 - 0)^2} = \sqrt{9 + 16} = \sqrt{25} = 5$
     
   - The Euclidean distance underestimates the true cost because it assumes diagonal movement, which is not allowed.

As far as computational complexity is concerned, the Euclidean distance requires squaring and taking the square root of coordinates, which is computationally more expensive than the Manhattan distance.
In scenarios with many states, this additional cost can significantly slow down heuristic evaluations.

For the previous reasons, from my point of view, the Manhattan distance is the better choice for grid-based problems with vertical and horizontal movements because:
1. It aligns perfectly with the movement constraints.
2. It is admissible and consistent, ensuring the correctness and efficiency of search algorithms like A*.
3. It is computationally simpler and directly reflects the true cost of moving between states.

The Euclidean distance, while useful in problems with free diagonal movement, is not suitable for grids with orthogonal movement due to its tendency to underestimate the actual cost and its higher computational cost.
As proof that it would be necessary to reconsider the choice of heuristic, I tried running the pathfinding program to solve a 5x5 puzzle instance and, despite letting it run for 260 minutes, I was unable to obtain a solution.

In conclusion, with optimizations and better documentation, this algorithm could become a robust and reusable solution.
